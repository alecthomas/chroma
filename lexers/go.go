package lexers

import (
	"strings"

	. "github.com/alecthomas/chroma" // nolint
)

// Go lexer.
var Go = Register(MustNewLexer(
	&Config{
		Name:      "Go",
		Aliases:   []string{"go", "golang"},
		Filenames: []string{"*.go"},
		MimeTypes: []string{"text/x-gosrc"},
	},
	Rules{
		"root": {
			{`\n`, Text, nil},
			{`\s+`, Text, nil},
			{`\\\n`, Text, nil},
			{`//(.*?)$`, CommentSingle, nil},
			{`/(\\\n)?[*](.|\n)*?[*](\\\n)?/`, CommentMultiline, nil},
			{`(import|package)\b`, KeywordNamespace, nil},
			{`(var|func|struct|map|chan|type|interface|const)\b`, KeywordDeclaration, nil},
			{Words(``, `\b`, `break`, `default`, `select`, `case`, `defer`, `go`, `else`, `goto`, `switch`, `fallthrough`, `if`, `range`, `continue`, `for`, `return`), Keyword, nil},
			{`(true|false|iota|nil)\b`, KeywordConstant, nil},
			{Words(``, `\b(\()`, `uint`, `uint8`, `uint16`, `uint32`, `uint64`, `int`, `int8`, `int16`, `int32`, `int64`, `float`, `float32`, `float64`, `complex64`, `complex128`, `byte`, `rune`, `string`, `bool`, `error`, `uintptr`, `print`, `println`, `panic`, `recover`, `close`, `complex`, `real`, `imag`, `len`, `cap`, `append`, `copy`, `delete`, `new`, `make`), ByGroups(NameBuiltin, Punctuation), nil},
			{Words(``, `\b`, `uint`, `uint8`, `uint16`, `uint32`, `uint64`, `int`, `int8`, `int16`, `int32`, `int64`, `float`, `float32`, `float64`, `complex64`, `complex128`, `byte`, `rune`, `string`, `bool`, `error`, `uintptr`), KeywordType, nil},
			{`\d+i`, LiteralNumber, nil},
			{`\d+\.\d*([Ee][-+]\d+)?i`, LiteralNumber, nil},
			{`\.\d+([Ee][-+]\d+)?i`, LiteralNumber, nil},
			{`\d+[Ee][-+]\d+i`, LiteralNumber, nil},
			{`\d+(\.\d+[eE][+\-]?\d+|\.\d*|[eE][+\-]?\d+)`, LiteralNumberFloat, nil},
			{`\.\d+([eE][+\-]?\d+)?`, LiteralNumberFloat, nil},
			{`0[0-7]+`, LiteralNumberOct, nil},
			{`0[xX][0-9a-fA-F]+`, LiteralNumberHex, nil},
			{`(0|[1-9][0-9]*)`, LiteralNumberInteger, nil},
			{`'(\\['"\\abfnrtv]|\\x[0-9a-fA-F]{2}|\\[0-7]{1,3}|\\u[0-9a-fA-F]{4}|\\U[0-9a-fA-F]{8}|[^\\])'`, LiteralStringChar, nil},
			{"`[^`]*`", LiteralString, nil},
			{`"(\\\\|\\"|[^"])*"`, LiteralString, nil},
			{`(<<=|>>=|<<|>>|<=|>=|&\^=|&\^|\+=|-=|\*=|/=|%=|&=|\|=|&&|\|\||<-|\+\+|--|==|!=|:=|\.\.\.|[+\-*/%&])`, Operator, nil},
			{`[|^<>=!()\[\]{}.,;:]`, Punctuation, nil},
			{`[^\W\d]\w*`, NameOther, nil},
		},
	},
).SetAnalyser(func(text string) float32 {
	if strings.Contains(text, "fmt.") && strings.Contains(text, "package ") {
		return 0.5
	}
	if strings.Contains(text, "package ") {
		return 0.1
	}
	return 0.0
}))
