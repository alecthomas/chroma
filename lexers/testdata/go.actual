package main

import "fmt"

func main() {
    fmt.Println("Hello World!")
}

var n int = 0x21 + 1_000
var n2 float64 = 1e3

func hello(a int) {
	fmt.Println("Hello World!").Hello()

    return func() int {
        return i
    }
} // One last thing