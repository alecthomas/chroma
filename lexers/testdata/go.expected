[
  {"type":"KeywordNamespace","value":"package"},
  {"type":"Text","value":" "},
  {"type":"NameOther","value":"main"},
  {"type":"Text","value":"\n\n"},
  {"type":"KeywordNamespace","value":"import"},
  {"type":"Text","value":" "},
  {"type":"LiteralString","value":"\"fmt\""},
  {"type":"Text","value":"\n\n"},
  {"type":"KeywordDeclaration","value":"func"},
  {"type":"Text","value":" "},
  {"type":"NameFunction","value":"main"},
  {"type":"Punctuation","value":"()"},
  {"type":"Text","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"Text","value":"\n    "},
  {"type":"NameOther","value":"fmt"},
  {"type":"Punctuation","value":"."},
  {"type":"NameFunction","value":"Println"},
  {"type":"Punctuation","value":"("},
  {"type":"LiteralString","value":"\"Hello World!\""},
  {"type":"Punctuation","value":")"},
  {"type":"Text","value":"\n"},
  {"type":"Punctuation","value":"}"},
  {"type":"Text","value":"\n\n"},
  {"type":"KeywordDeclaration","value":"var"},
  {"type":"Text","value":" "},
  {"type":"NameOther","value":"n"},
  {"type":"Text","value":" "},
  {"type":"KeywordType","value":"int"},
  {"type":"Text","value":" "},
  {"type":"Punctuation","value":"="},
  {"type":"Text","value":" "},
  {"type":"LiteralNumberHex","value":"0x21"},
  {"type":"Text","value":" "},
  {"type":"Operator","value":"+"},
  {"type":"Text","value":" "},
  {"type":"LiteralNumberInteger","value":"1_000"},
  {"type":"Text","value":"\n"},
  {"type":"KeywordDeclaration","value":"var"},
  {"type":"Text","value":" "},
  {"type":"NameOther","value":"n2"},
  {"type":"Text","value":" "},
  {"type":"KeywordType","value":"float64"},
  {"type":"Text","value":" "},
  {"type":"Punctuation","value":"="},
  {"type":"Text","value":" "},
  {"type":"LiteralNumberFloat","value":"1e3"},
  {"type":"Text","value":"\n\n"},
  {"type":"KeywordDeclaration","value":"func"},
  {"type":"Text","value":" "},
  {"type":"NameFunction","value":"hello"},
  {"type":"Punctuation","value":"("},
  {"type":"NameOther","value":"a"},
  {"type":"Text","value":" "},
  {"type":"KeywordType","value":"int"},
  {"type":"Punctuation","value":")"},
  {"type":"Text","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"Text","value":"\n\t"},
  {"type":"NameOther","value":"fmt"},
  {"type":"Punctuation","value":"."},
  {"type":"NameFunction","value":"Println"},
  {"type":"Punctuation","value":"("},
  {"type":"LiteralString","value":"\"Hello World!\""},
  {"type":"Punctuation","value":")."},
  {"type":"NameFunction","value":"Hello"},
  {"type":"Punctuation","value":"()"},
  {"type":"Text","value":"\n\n    "},
  {"type":"Keyword","value":"return"},
  {"type":"Text","value":" "},
  {"type":"KeywordDeclaration","value":"func"},
  {"type":"Punctuation","value":"()"},
  {"type":"Text","value":" "},
  {"type":"KeywordType","value":"int"},
  {"type":"Text","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"Text","value":"\n        "},
  {"type":"Keyword","value":"return"},
  {"type":"Text","value":" "},
  {"type":"NameOther","value":"i"},
  {"type":"Text","value":"\n    "},
  {"type":"Punctuation","value":"}"},
  {"type":"Text","value":"\n"},
  {"type":"Punctuation","value":"}"},
  {"type":"Text","value":" "},
  {"type":"CommentSingle","value":"// One last thing\n"}
]
